Add text on hash in KB-JWT (#353)
oauth-wg/oauth-selective-disclosure-jwt: draft-ietf-oauth-selective-disclosure-jwt.md

Note: How the public key is included in SD-JWT is out of scope of this document. It can be passed by value or by reference.

For presenting an SD-JWT to a Verifier that enforces Key Binding, the Holder
creates a signed document, the Key Binding JWT as defined in (#kb-jwt), using

> its private key. This document contains some data provided by the Verifier, such
> as a nonce, to ensure the freshness of the signature, an audience value to
> indicate the intended audience for the document, and a hash that ensures the
> integrity of the data sent to the Verifier. Details of the format of Key Binding JWTs are

Reviewer: The hash isn't provided by the Verifier so suggest breaking the hash part out of the sentence that starts "This document contains some data provided by the Verifier".  Also be a bit more specific that it's integrity of the data that the holder sent to the Verifier.  Something like: 

```suggestion
its private key. This document contains some data provided by the Verifier, such
as a nonce, to ensure the freshness of the signature, and an audience value to
indicate the intended audience for the document. It also includes a hash that ensures the
integrity of the data sent from the Holder to the Verifier. Details of the format of Key Binding JWTs are
```

Author: But the audience value is also not "provided" by the Verifier... so....
```suggestion
its private key. This document usually contains a nonce to ensure the freshness of the signature, an audience value to
indicate the intended audience for the document, and a hash that ensures the
integrity of the data sent to the Verifier. Details of the format of Key Binding JWTs are
```

Reviewer: Yeah, the audience value isn't "provided" by the Verifier exactly but it could be from the verifier's metadata or HTTPS cert or something so does kinda come from the verifier. Also "provided" was used in the text prior to this PR so I was just keeping with that. But you are right (as you usually are!). Here's some proposed text riffing on your text that drops "usually" because the KB-JWT is defined to have these things and is a bit more specific that it's integrity of the data that the holder sent to the Verifier.

```suggestion
its private key. This document contains a nonce to ensure the freshness of the signature, an audience value to
indicate the intended audience for the document, and a hash that ensures the
integrity of the data sent from the Holder to the Verifier. Details of the format of Key Binding JWTs are
```

    * `iat`: REQUIRED. The value of this claim MUST be the time at which the Key Binding JWT was issued using the syntax defined in [@!RFC7519].
    * `aud`: REQUIRED. The intended receiver of the Key Binding JWT. How the value is represented is up to the protocol used and out of scope of this specification.
    * `nonce`: REQUIRED. Ensures the freshness of the signature. The value type of this claim MUST be a string. How this value is obtained is up to the protocol used and out of scope of this specification.
    * `_sd_hash`: REQUIRED. The base64url-encoded hash digest over the Issuer-signed JWT and the selected Disclosures as defined below.

> ### Integrity Protection of the Presentation

Author: ```suggestion
### Integrity Protection of the Presentation {#integrity-protection-of-the-presentation}
```

       5. Check that the `typ` of the Key Binding JWT is `kb+jwt`.
       6. Check that the creation time of the Key Binding JWT, as determined by the `iat` claim, is within an acceptable window.
       7. Determine that the Key Binding JWT is bound to the current transaction and was created for this Verifier (replay protection) by validating `nonce` and `aud` claims.

>        8. Calculate the digest over the Presentation and verify that it matches the value of the `_sd_hash` claim in the Key Binding JWT.

Reviewer 2: I think (but am not entirely sure) that "Presentation" includes the key binding jwt, but I think the hash should exclude the key binding jwt? ('Presentation' doesn't seem to be a defined term, I opened https://github.com/oauth-wg/oauth-selective-disclosure-jwt/issues/356 )

Reviewer 2: Suggestion looks good to me.

In some scenarios this behavior could be desirable,
but if it is not, Issuers need to support and Verifiers need to enforce Key Binding.

## Integrity of Presentation

In a Presentation, the Issuer-signed JWT is integrity-protected by the Issuer's
signature, and the Disclosures are integrity-protected by the digests included

> in the Issuer-signed JWT. However, without enforcing a Key Binding JWT as
> defined in (#kb-jwt), the integrity of the set of Disclosures the

Reviewer 3: ```suggestion
in the Issuer-signed JWT. However, without enforcing the digest
over the Issuer-signed JWT and Disclosures as defined in (#integrity-protection-of-the-presentation)
in the Key Binding JWT as defined in (#kb-jwt), the integrity of the set of Disclosures the
```

> Holder selected to send is not ensured.

Reviewer 4: ```suggestion
Holder disclosed is not ensured.
```

Reviewer 4: I might rework the phrasing of this section,

The specific set of disclosures presented MAY be integrity protected through the use of a hash or other means, etc... it reads a bit awkward as is

In some scenarios this behavior could be desirable,
but if it is not, Issuers need to support and Verifiers need to enforce Key Binding.

## Integrity of Presentation

In a Presentation, the Issuer-signed JWT is integrity-protected by the Issuer's
signature, and the Disclosures are integrity-protected by the digests included

> in the Issuer-signed JWT. However, without enforcing a Key Binding JWT as
> defined in (#kb-jwt), an attacker that can intercept the Presentation can remove
> Disclosures. The attacker could in theory also add Disclosures, but that would
> require the attacker to know the salt and claim value for the claim to be added,
> which is infeasible if the salt has sufficient entropy.

Reviewer: I am sorry for getting hung up on this one but I do worry that talking about "an attacker that can intercept the Presentation" here might cause a lot of questions/consternation/concern from readers and reviewers. So here's an attempt at saying more or less the same thing but with fewer words and no specific mention of an attacker. 

```suggestion
in the Issuer-signed JWT. However, without enforcing a Key Binding JWT as
defined in (#kb-jwt), the integrity of the selected set of Disclosures the 
Holder elected to send is not ensured.
```

Author: elected or selected?

Reviewer: The word "selected" was already in the sentence earlier so I went with "elected" for word choice variety. But "selected" would be fine, if you think it's better/clearer?

Reviewer: or ... change so that "selected" is used just once? 

```suggestion
in the Issuer-signed JWT. However, without enforcing a Key Binding JWT as
defined in (#kb-jwt), the integrity of the set of Disclosures the 
Holder selected to send is not ensured.
```
